mkosi/gentoo.py

```diff
@@ -0,0 +1,510 @@
+
+
+# https://github.com/python/mypy/issues/5732
+kimg_path: str
+
+
+class Gentoo:
+    arch_profile: str
+    emerge_default_opts: Set[str]
+    EMERGE_UPDATE_OPTS: Set[str]
+    arch: str
+    package_use: str
+    pkgs_boot: List[Set[str]]
+    pkgs_sys: Set[str]
+    pkgs_fs: Set[str]
+    grub_platforms: Set[str]
+    portage_features: Set[str]
+    portage_use_flags: Set[str]
+    UNINSTALL_IGNORE: Set[str]
+    dracut_atom = "sys-kernel/dracut"
+
+    def __init__(self, args: CommandLineArguments, root: str, do_run_build_script: bool) -> None:
+        self.portage_cfg = config(config_root=root, target_root=root, sysroot=root, eprefix=None)
+        if "build-script" in ARG_DEBUG:
+            for c in self.portage_cfg:
+                print("%-32.32s = %-s" % (c, self.portage_cfg[c]))
+
+        jobs = multiprocessing.cpu_count()
+        self.emerge_default_opts = {
+            "--buildpkg=y",
+            "--usepkg=y",
+            "--keep-going=y",
+            "--jobs=" + str(jobs),
+            "--load-average=" + str(jobs - 1),
+            "--nospinner",
+        }
+        if "build-script" in ARG_DEBUG:
+            self.emerge_default_opts |= {"--verbose"}
+            self.emerge_default_opts |= {"--quiet=n"}
+            self.emerge_default_opts |= {"--quiet-fail=n"}
+        else:
+            self.emerge_default_opts |= {"--quiet-build"}
+            self.emerge_default_opts |= {"--quiet"}
+
+        self.EMERGE_UPDATE_OPTS = {
+            "--update",
+            "--tree",
+            "--changed-use",
+            "--newuse",
+            "--deep",
+            "--with-bdeps=y",
+            "--complete-graph-if-new-use=y",
+        }
+
+        # 'systemd' is hard dependancy?
+        # 'git' for sync-type=git
+        # 'symlink' for kernel
+        self.portage_use_flags = {
+            "systemd",
+            "initramfs",
+            "git",
+            "symlink",
+            "sdl",
+            "-filecaps",
+            "-savedconfig",
+            "-split-bin",
+            "-split-sbin",
+            "-split-usr",
+        }
+
+        # -user* are required for access to USER_CONFIG_PATH
+        # -pid-sandbox is required for cross compile scenarios
+        self.portage_features = {
+            "-userfetch",
+            "-userpriv",
+            "-usersync",
+            "-usersandbox",
+            "-sandbox",
+            "-pid-sandbox",
+            "-network-sandbox",
+            "parallel-install",
+            "buildpkg",
+            "binpkg-multi-instance",
+            "-binpkg-docompress",
+            "getbinpkg",
+            "-candy",
+        }
+        # TODO: portage_features.add("ccache")
+        os.environ["FEATURES"] = " ".join(self.portage_features)
+        os.environ["BOOTSTRAP_USE"] = " ".join(self.portage_use_flags)
+        os.environ["USE"] = " ".join(self.portage_use_flags)
+        os.environ["EGIT_CLONE_TYPE"] = "shallow"
+
```

Comment:
Those should be passed to the invoked process. Don't modify the environment.

Reply:
> This approach seams much better then the previous version. When it's all in a separate file, it doesn't seem so bad ;)
> 
> Unfortunately we changed the types of all path-like variables to `Path` in main, so this patch will need to be reworked :(
> I don't know gentoo at all, so some comments about python only.

updated to `Path`, please review and let me know if i missed something.

Reply:
i thinkg i removed all the refs to `os.envion`. please verify.